reference/method-patterns.md

```diff
@@ -29,19 +29,27 @@ The return type of a method is _not_ represented in method patterns. Methods in
+| `java.lang.String substring(int, int)`      | Exactly the [two argument overload](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/String.html#substring%28int,int%29) of `String.substring(int beginIndex, int endIndex)`              |
+| `java.lang.String substring(..)`            | Any overload of `String.substring()` with any number of arguments                                                                                                                                               |
+| `java.lang.String *(int)`                   | Any method on `String` that accepts a single argument of type `int`                                                                                                                                             |
+| `java.lang.String format(String, ..)`       | Exactly the [String.format(String format, Object... args)](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/String.html#format(java.lang.String,java.lang.Object...)) method with varargs |
+| `java.lang.String format(String, Object[])` | Exactly the [String.format(String format, Object... args)](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/String.html#format(java.lang.String,java.lang.Object...)) method with varargs |
+| `com.yourorg.Foo bar(int, String, ..)`      | Any method on `Foo` named `bar` accepting an `int`, a `String`, and zero or more other arguments of any type                                                                                                    |
+| `com.yourorg.Foo <constructor>(*, *, *)`    | Constructors of `Foo` accepting exactly three arguments of any type                                                                                                                                             |
+| `*..String *(..)`                           | Any method accepting any arguments on classes named "String" in any package                                                                                                                                     |
+| `*..* *(..)`                                | Any method accepting any arguments on any class                                                                                                                                                                 |
+| `org.example..* *(..)`                      | Any method on any class in the "org.example" package, or any sub-package of "org.example"                                                                                                                       |
+
+{% hint style="info" %}
+When matching methods using varags, method patterns should match the parameter _declaration_, not the parameter _use_ for method invocations.
```

Comment:
Figured an extra comment on matching varargs couldn't hurt. Didn't fit well into the table above.

Reply:
Might even expand this with something like:

A method can be declared as `org.apache.commons.lang3.Validate notNull(Object, String, Object[])`, but can be invoked with a variable number of arguments.
```
Validate.notNull(obj, "Obj was null")
Validate.notNull(obj, "Obj %s was null", "foo")
Validate.notNull(obj, "Obj %s null in %", "foo", "bar")
```
When matching any of the above method invocations, use the method declaration rather than the actual number of arguments.

Reply:
@timtebeek this hint is a bit misleading. `MethodMatcher` _always_ match against declared parameters, not supplied parameters. You're matching the method _being_ invoked, not the method invocation itself, if that makes sense.